File: #86_Partition_List/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class Solution {
    public ListNode partition(ListNode head, int x) {
        if (head == null) return head;
        ListNode less = new ListNode(0);
        ListNode greater = new ListNode(0);
        ListNode curLess = less;
        ListNode curGreater = greater;
        ListNode cur = head;
        while (cur != null) {
        	if (cur.val < x) {
        		curLess.next = cur;
        		curLess = curLess.next;
        	}
        	else {
        		curGreater.next = cur;
        		curGreater = curGreater.next;
        	}
        	cur = cur.next;
        }
        curLess.next = greater.next;
        curGreater.next = null;
        return less.next;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	ListNode n1 = new ListNode(1);
    	ListNode n2 = new ListNode(4);
    	ListNode n3 = new ListNode(3);
    	ListNode n4 = new ListNode(2);
    	ListNode n5 = new ListNode(5);
    	ListNode n6 = new ListNode(2);
    	n1.next = n2;
    	n2.next = n3;
    	n3.next = n4;
    	n4.next = n5;
    	n5.next = n6;
    	ListNode nd = sol.partition(n1, 3);
    	while(nd != null) {
    		System.out.print(nd.val + " ");
    		nd = nd.next;
    	}
    	System.out.println();
    }
}

File: #86_Partition_List/Solution.py
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None

class Solution:
	# @param {ListNode} head
	# @param {integer} x
	# @return {ListNode}
	def partition(self, head, x):
		left = ListNode(0)
		right = ListNode(0)
		leftCur = left
		rightCur = right
		cur = head
		while cur:
			if cur.val < x:
				leftCur.next = cur
				leftCur = leftCur.next
			else:
				rightCur.next = cur
				rightCur = rightCur.next
			cur = cur.next
			leftCur.next = None
			rightCur.next = None
		leftCur.next = right.next
		right.next = None
		return left.next